traits/ctraits.c

```diff
@@ -3677,8 +3677,9 @@ validate_trait_callable(
     trait_object *trait, has_traits_object *obj, PyObject *name,
     PyObject *value)
 {
+    int allow_none = PyObject_IsTrue(PyTuple_GET_ITEM(trait->py_validate, 1));
 
-    if ((value == Py_None) || PyCallable_Check(value)) {
+    if ((allow_none && value == Py_None) || PyCallable_Check(value)) {
```

Comment:
This is something of a micro-optimisation, but I think we should only unpack and interpret the second element of `trait->py_validate` after we've checked `value == Py_None`: the `== Py_None` check is cheap, but the `PyObject_IsTrue` call will be more time-consuming.

We also need to catch the case where `PyObject_IsTrue` returns `-1`: in that case, something unexpected has happened and we should be returning `NULL` immediately.